File: app/src/main/java/com/doapps/habits/adapter/HabitRecycleAdapter.java
package com.doapps.habits.adapter;

import android.graphics.Color;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.doapps.habits.R;
import com.doapps.habits.listeners.EmptyListListener;
import com.doapps.habits.models.Habit;
import com.doapps.habits.models.IHabitDatabaseMovableListProvider;
import com.doapps.habits.models.IHabitsDatabase;
import com.doapps.habits.models.IMovableList;
import com.doapps.habits.viewholder.HabitViewHolder;

import java.util.Calendar;
import java.util.List;

public class HabitRecycleAdapter extends RecyclerView.Adapter<HabitViewHolder> implements IMovableList {
    private final List<Habit> mHabitList;
    private final IHabitsDatabase mHabitsDatabase;
    private final IHabitDatabaseMovableListProvider mMovableList;

    public HabitRecycleAdapter(IHabitDatabaseMovableListProvider movableList) {
        mMovableList = movableList;
        mHabitList = movableList.getList();
        mHabitsDatabase = movableList.getDatabase();
    }

    @Override
    public HabitViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view =
                LayoutInflater.from(parent.getContext()).inflate(R.layout.habit_list_item, parent, false);
        return new HabitViewHolder(view);
    }

    @Override
    public void onBindViewHolder(HabitViewHolder holder, int position) {
        Log.d("HabitAdapter", "Position is: " + position);
        Habit habit = mHabitList.get(position);

        holder.txtTitle.setText(habit.title);
        holder.checkBox.setOnClickListener(view -> {
            if (holder.checkBox.isChecked()) {
                mHabitsDatabase.updateHabit(habit, 1);
                mHabitList.get(position).setDoneMarker(true);
                holder.txtTitle.setTextColor(Color.GRAY);
            } else {
                mHabitsDatabase.updateHabit(habit, 0);
                mHabitList.get(position).setDoneMarker(false);
                holder.txtTitle.setTextColor(Color.BLACK);
            }
        });

        Calendar calendar = Calendar.getInstance();

        holder.checkBox.setChecked(habit.isDone(
                calendar.get(Calendar.DATE),
                calendar.get(Calendar.MONTH),
                calendar.get(Calendar.YEAR)));

    }

    @Override
    public int getItemCount() {
        return mHabitList.size();
    }

    @Override
    public void onItemDismiss(int position) {
        mMovableList.onItemDismiss(position);
        notifyItemRemoved(position);
        EmptyListListener.listener.isEmpty(getItemCount() == 0);
    }

    @Override
    public void onItemMove(int fromPosition, int toPosition) {
        mMovableList.onItemMove(fromPosition, toPosition);
        notifyItemMoved(fromPosition, toPosition);
    }
}

File: app/src/main/java/com/doapps/habits/adapter/HabitRecycleAdapter.kt
package com.doapps.habits.adapter

import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.doapps.habits.R
import com.doapps.habits.helper.HabitListManager
import com.doapps.habits.listeners.EmptyListListener
import com.doapps.habits.models.Habit
import com.doapps.habits.view.holders.HabitViewHolder
import java.util.*

class HabitRecycleAdapter(private val movableHabitList: HabitListManager)
  : RecyclerView.Adapter<HabitViewHolder>(), IMovableListAdapter {

  /**
   * A list of [com.doapps.habits.models.Habit]s for binding
   */
  private val habitList: MutableList<Habit> = movableHabitList.list

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HabitViewHolder {
    val view = LayoutInflater.from(parent.context).inflate(R.layout.habit_list_item, parent, false)
    return HabitViewHolder(view)
  }

  override fun onBindViewHolder(holder: HabitViewHolder, position: Int) {
    val habit = habitList[position]

    holder.titleTextView.text = habit.title
    holder.checkBox.setOnClickListener {
      habit.isDoneMarker = holder.checkBox.isChecked
      movableHabitList.update(habit)
      holder.titleTextView.setTextColor(if (habit.isDoneMarker) Color.GRAY else Color.BLACK)
      holder.setProgressBar(habit)
    }
    holder.titleTextView.setTextColor(if (habit.isDoneMarker) Color.GRAY else Color.BLACK)
    holder.setProgressBar(habit)

    val calendar = Calendar.getInstance()

    holder.checkBox.isChecked = habit.isDone(
        calendar.get(Calendar.DAY_OF_MONTH),
        calendar.get(Calendar.MONTH),
        calendar.get(Calendar.YEAR))

  }

  /**
   * The function to return number of items in list
   * @return Number of items in list
   */
  override fun getItemCount() = habitList.size

  /**
   * Handles item dismissing
   * @param position item position to be dismissed
   */
  override fun onItemDismiss(position: Int) {
    movableHabitList.onItemDismiss(position)
    habitList.removeAt(position)
    EmptyListListener.listener.isEmpty(habitList.size == 0)
    notifyItemRemoved(position)
  }

  /**
   * The function to handle item moving in list
   * @param fromPosition Item start position
   * @param toPosition Item finish position
   */
  override fun onItemMove(fromPosition: Int, toPosition: Int) {
    Collections.swap(habitList, fromPosition, toPosition)
    val fromPositionId = habitList[fromPosition].id
    habitList[fromPosition].id = habitList[toPosition].id
    habitList[toPosition].id = fromPositionId
    movableHabitList.onItemMove(habitList[fromPosition], habitList[toPosition])
    notifyItemMoved(fromPosition, toPosition)
  }
}

